### PR TITLE
Allow _key style wildcard patterns.

### PR DESCRIPTION
**Pull request by allanderek**
First attempt at allowing wildcard patterns '_' to have a variable na…me after, whih is intended to allow documentation of the value that is being ignored: https://github.com/gren-lang/compiler/issues/111

### CODE REVIEW

File: tests/Parse/UnderscorePatternSpec.hs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+data ParseError
+  = ExprError P.Row P.Col
+  | OtherError String P.Row P.Col
+  deriving (Show, Eq)
+
+spec :: Spec
+spec = do
+  Hspec.describe "Wildcard patterns" $ do
+    Hspec.it "regression test" $
```
Comment (robinheghan) [inline tests/Parse/UnderscorePatternSpec.hs:22]: Would you mind importing `describe` and `it` unqualified, as well as adding a line of whitespace between each test?

Just so the formatting is the same between all tests.

**Review by robinheghan — COMMENTED** [review]
If I'm understanding this correctly, this simply allows leading underscores in variable names. A future PR will make sure no such variable is referenced. Is that right?

If so, I'm going to hold back on merging the PR until the second part lands. I want to avoid having to revert this if no one has time to do that second PR.

As for the code itself: nice work! The tests are good, there's a good descriptive comment and the `isReservedWord` function helps with readability :)

**Comment by allanderek** [comment]
> If I'm understanding this correctly, this simply allows leading underscores in variable names. A future PR will make sure no such variable is referenced. Is that right?

Not quite. It only allows this in **patterns**, not expressions. So you cannot do:

```elm
myFunction : a -> a
myFunction _x = _x
```
In addition, the 'x' part of the pattern is not saved as part of the wild card AST: https://github.com/gren-lang/compiler/blob/main/compiler/src/AST/Source.hs#L95 (unchanged by the commit)
So even if the parser allowed the above `_x` in the body of the function, it wouldn't be in scope.

However, I don't know how gren-format is working? Does it use the compiler's parser? If so, then it's going to remove the `x` part of the `_x` pattern.

In addition, you might want to give a really good error message in the case that someone **does** attempt to use such a variable. The expression parser could be modified to explain the situation if you attempted to use `_x` as a variable expression.

**Comment by robinheghan** [comment]
> It only allows this in patterns, not expressions.

Ahh. Are you planning a follow up PR?

> However, I don't know how gren-format is working? Does it use the compiler's parser?

It does, yeah. So the way this is implemented would make gren-format remove such statements from the code 😅

**Comment by allanderek** [comment]
> > It only allows this in patterns, not expressions.
> 
> Ahh. Are you planning a follow up PR?

To allow for this in expressions? No, it seems to me a good thing that it's not allowed in expressions. I think there is a slight confusion, you seem to think that I've done this by changing how lower-cased identifiers are parsed. Instead, I've done this by changing how the wildcard `_` pattern is parsed, it now allows there to be an optional lower-cased identifier suffix. But that suffix is currently simply thrown away.

> > However, I don't know how gren-format is working? Does it use the compiler's parser?
> 
> It does, yeah. So the way this is implemented would make gren-format remove such statements from the code sweat_smile

However, I could change the parser so that the AST for wildcard patterns contains the lower-case identifier suffix. If I do this, I believe it would break gren-format, because I would be changing the type of the wild card pattern. However, it would be broken in a way that it would be forced to be fixed, and the fix, I would think, would be relatively straightforward.

Should I go ahead and make that change? I can just keep such a change in this PR. I guess I should have used a branch, but I'm not planning on doing anything else before this is finished.

The other thing I could add would be to try to give a nicer error message in the case that one attempts to use `_x` as an **expression**. I'm a little less confident about being able to test for that well though.

**Comment by robinheghan** [comment]
> I think there is a slight confusion

I read your example as if you wouldn't be able to do `func _x = ...`, but I see now that you were refering to not being able to use `_x` as an expression in the body 🤦‍♂️ 

> I could change the parser so that the AST for wildcard patterns contains the lower-case identifier suffix

Please do, @avh4 or me will fix whatever needs fixing in gren format.

> The other thing I could add would be to try to give a nicer error message

If you have the time and energy, I would greatly appreciate that. Not entirely sure how to test for it, either, but I'm sure we can figure it out.

By the way, would you mind adding your name to the `CONTRIBUTORS` file?

**Comment by allanderek** [comment]
> > I could change the parser so that the AST for wildcard patterns contains the lower-case identifier suffix
> 
> Please do, @avh4 or me will fix whatever needs fixing in gren format.

Okay will do

> > The other thing I could add would be to try to give a nicer error message
> 
> If you have the time and energy, I would greatly appreciate that. Not entirely sure how to test for it, either, but I'm sure we can figure it out.
> 

I'll try.

> By the way, would you mind adding your name to the `CONTRIBUTORS` file?

Sure no problem.

I also saw that my commits are failing due to not being formatted by `ormolu`, which I will fix. Is there someway I can run the set of verifiers locally?

**Comment by allanderek** [comment]
Okay I've got something that gives a particular error message in the case that the user attempts to use an identifier that begins with an underscore:
```

> let
|   my_function _allan = _allan
| in
| my_function 10
|
-- ATTEMPT TO USE WILDCARD VARIABLE --------------------------------------- REPL

It appears you are try to use a variable name that starts with an underscore
(_allan). Such variable names can appear in patterns but not expressions. Such a
pattern is equivalent to using a single '_' pattern with some documentation.

4|     my_function _allan = _allan
                                  ^
Perhaps rename the variable without the underscore prefix.
```

I definitely think the message could do with some work, and there is no test for this yet.
I'm out at the weekend though so I'm not sure when I'll have a chance to look at this again.

**Comment by allanderek** [comment]
Okay I've managed to add a test for the case in which the user attempts to use a `_key` style variable in an expression.
To do this I added a couple of general parser test helpers and I think this makes the whole module for testing underscore patterns neater.

I think this is now ready.

I also think that updating gren-format to account for this will be straightforward. The AST node for a wildcard pattern `PAnything` now just has an argument `PAnything Name.Name`. So I now think it has everything to print out such patterns correctly, or calculate how long they are, or whatever is needed.

**Comment by robinheghan** [comment]
Looks good to me!

Thank you for this, Allan 💯

**Pull request merged**